src/crimson/net/chained_dispatchers.cc

```diff
@@ -13,19 +13,27 @@ seastar::future<>
 ChainedDispatchers::ms_dispatch(crimson::net::Connection* conn,
                                 MessageRef m) {
   try {
-    return seastar::do_for_each(dispatchers, [conn, m](Dispatcher& dispatcher) {
-      return dispatcher.ms_dispatch(conn, m);
-    }).handle_exception([conn] (std::exception_ptr eptr) {
-      logger().error("{} got unexpected exception in ms_dispatch() throttling {}",
-                     *conn, eptr);
-      ceph_abort();
-    });
+    for (auto& dispatcher : dispatchers) {
+      auto [dispatched, throttle_future] = dispatcher.ms_dispatch(conn, m);
+      if (dispatched) {
+        return std::move(throttle_future
+        ).handle_exception([conn] (std::exception_ptr eptr) {
+          logger().error("{} got unexpected exception in ms_dispatch() throttling {}",
+                         *conn, eptr);
+          ceph_abort();
+        });
+      }
+      assert(throttle_future.available());
+    }
   } catch (...) {
     logger().error("{} got unexpected exception in ms_dispatch() {}",
                    *conn, std::current_exception());
     ceph_abort();
-    return seastar::now();
   }
+  if (!dispatchers.empty()) {
```

Comment:
does this really happen in real world? i mean it's unlikely to start pumping messages before a messenger is connected with dispatchers.

Reply:
`Messenger::stop()` is implemented by clearing the `dispatchers`, so that event dispatching can be stopped immediately.

Reply:
so, in that case, `dispatchers` is empty, right?

Reply:
Yep, see https://github.com/ceph/ceph/blob/659420c64cd379f1cb95e4b0a17113c3ddfea270/src/crimson/net/SocketMessenger.h#L83-L85 `is_started()` is also implemented by checking if dispatchers is empty. A started Messenger cannot contain empty dispatchers.

Reply:
so, i think probably we can just `assert(dispatchers.empty())` here. what do you think?

Reply:
An error message here is necessary to show which message is not handled. I think it may be too strict to crash OSD for an unhandled message because it may not be the fault of this Crimson-OSD. What do you think?

Reply:
i see, a non-empty `dispatchers` signifies a TODO actually. as we either 

- fail to handle a message, or
- fail to stop sending a message which is not used anymore